easybuild/framework/easyconfig/default.py

```diff
@@ -146,6 +146,7 @@
     'builddependencies': [[], "List of build dependencies", DEPENDENCIES],
     'dependencies': [[], "List of dependencies", DEPENDENCIES],
     'hiddendependencies': [[], "List of dependencies available as hidden modules", DEPENDENCIES],
+    'multi_deps': [{}, "Dict of lists of dependency versions over which to iterate", DEPENDENCIES],
```

Comment:
I know that this only makes for the case of build_deps but I still think that there's no harm in making that connection explicit here (for those who haven't thought so  deeply about it), how about `multi_builddeps`?

Reply:
I guess that could make sense, but it's kind of double...

What we're doing is installing something for multiple versions of a particular dependency (e.g. `Python`) in a single installation prefix.
Those dependencies are generally *runtime* dependencies, but during the installation process we only use them as build dependencies, to avoid that they are included as a runtime dep in the generated module file.
That's the only way we can iterate over dependencies without introducing conflicts in the generated module file.
That this is done by injecting stuff into `builddependencies` doesn't have to be explicit imho, it's kind of an implementation detail...

One concern with the multi-Python installations (raised by @Micket) is that users will have to know to load a Python module as well for things to work.
This could be remedied by loading one of the `Python` versions (e.g. the first listed in `multi_deps`) by default, if no other `Python` versions is loaded already (perhaps with an accompanying message). The logical choice there would be the Python 3.x version. Note that this still leaves the door open to use another `Python` version, which is just a `module swap` away...

I haven't made up my mind how to implement that yet (and it should probably be part of this PR)...
It could be done in general in the framework, or only for Python packages (but then we'll need to touch multiple generic easyblock like `PythonPackage`, `PythonBundle`, etc.).

Reply:
Ok, that makes sense.

Regarding letting users know a Python module is required is something that can be handled with `depends_on('Python')` (for Lmod)  without providing a version. That way the default is chosen by the site, without them doing something explicit it would be the latest version of Python 3. This could also work in general, `depends_on('%(name)s')`. Of course it would mean that if they had multiple subversions of Python the dep tree might change.

Reply:
@ocaisa That won't work imho, since you'd be in trouble when a new `Python` 3.x module is installed with a different subtoolchain...

Reply:
There you go with your flat naming scheme ruining the party again!

Reply:
Support for loading a default module is not implemented yet in this PR, I'll do that in a follow-up PR to avoid making this PR too big/complicated...

It's a nice enhancement to have, but not a strict requirement to make this work...

Reply:
+1 to the `depends_on` support in the future. One could have a guard, to check if the MNS is hierarchical or not, and use `depends_on` selectively on those cases. But the guard might not be trivial to make it general for custom MNS